File: src/main/java/com/example/dalattravel/service/MinCostMaxFlowService.java
package com.example.dalattravel.service;

import com.example.dalattravel.model.Passenger;
import com.example.dalattravel.model.PassengerGroup;
import com.example.dalattravel.model.Vehicle;
import lombok.Getter;
import lombok.Setter;
import org.springframework.stereotype.Service;

import java.math.BigDecimal;
import java.util.*;
import java.util.stream.Collectors;

@Service
public class MinCostMaxFlowService {

    @Getter
    @Setter
    public static class FlowEdge {
        private int from;
        private int to;
        private int capacity;
        private BigDecimal cost;
        private int flow;
        private FlowEdge reverse;
    }

    @Getter
    @Setter
    public static class FlowResult {
        private int maxFlow;
        private BigDecimal minCost = BigDecimal.ZERO;
        private Map<Integer, List<Integer>> assignments = new HashMap<>(); // VehicleId -> List of PassengerIds
    }

    public FlowResult solve(List<Vehicle> vehicles, List<Passenger> passengers, List<PassengerGroup> groups) {
        int source = 0;
        int sink = 1;
        int vehicleStart = 2;
        int passengerStart = vehicleStart + vehicles.size();
        int groupStart = passengerStart + passengers.size();

        Map<Integer, List<FlowEdge>> graph = new HashMap<>();
        List<FlowEdge> allEdges = new ArrayList<>();

        int totalNodes = groupStart + groups.size() + 2;
        for (int i = 0; i < totalNodes; i++) {
            graph.put(i, new ArrayList<>());
        }

        // Source -> Vehicles
        for (int i = 0; i < vehicles.size(); i++) {
            int vehicleNode = vehicleStart + i;
            FlowEdge edge = createEdge(source, vehicleNode, vehicles.get(i).getAvailableSeats(), BigDecimal.ZERO);
            graph.get(source).add(edge);
            graph.get(vehicleNode).add(edge.getReverse());
            allEdges.add(edge);
        }

        // Vehicles -> Passengers
        for (int v = 0; v < vehicles.size(); v++) {
            int vehicleNode = vehicleStart + v;
            Vehicle vehicle = vehicles.get(v);

            for (int p = 0; p < passengers.size(); p++) {
                Passenger passenger = passengers.get(p);
                if (passenger.getGroupId() != null) continue;

                int passengerNode = passengerStart + p;
                BigDecimal cost = calculateCost(vehicle, passenger);
                FlowEdge edge = createEdge(vehicleNode, passengerNode, 1, cost);
                graph.get(vehicleNode).add(edge);
                graph.get(passengerNode).add(edge.getReverse());
                allEdges.add(edge);
            }
        }

        // Vehicles -> Groups
        for (int v = 0; v < vehicles.size(); v++) {
            int vehicleNode = vehicleStart + v;
            Vehicle vehicle = vehicles.get(v);

            for (int g = 0; g < groups.size(); g++) {
                PassengerGroup group = groups.get(g);
                if (vehicle.getAvailableSeats() < group.getRequiredSeats()) continue;

                int groupNode = groupStart + g;
                BigDecimal cost = calculateGroupCost(vehicle, group);
                FlowEdge edge = createEdge(vehicleNode, groupNode, 1, cost);
                graph.get(vehicleNode).add(edge);
                graph.get(groupNode).add(edge.getReverse());
                allEdges.add(edge);
            }
        }

        // Groups -> Group Passengers
        for (int g = 0; g < groups.size(); g++) {
            int groupNode = groupStart + g;
            PassengerGroup group = groups.get(g);
            List<Passenger> groupPassengers = passengers.stream()
                    .filter(p -> Objects.equals(p.getGroupId(), group.getId()))
                    .collect(Collectors.toList());

            for (Passenger passenger : groupPassengers) {
                int passengerNode = passengerStart + passengers.indexOf(passenger);
                FlowEdge edge = createEdge(groupNode, passengerNode, 1, BigDecimal.ZERO);
                graph.get(groupNode).add(edge);
                graph.get(passengerNode).add(edge.getReverse());
                allEdges.add(edge);
            }
        }

        // Passengers -> Sink
        for (int p = 0; p < passengers.size(); p++) {
            int passengerNode = passengerStart + p;
            FlowEdge edge = createEdge(passengerNode, sink, 1, BigDecimal.ZERO);
            graph.get(passengerNode).add(edge);
            graph.get(sink).add(edge.getReverse());
            allEdges.add(edge);
        }

        return minCostMaxFlow(graph, source, sink, vehicleStart, passengerStart, groupStart, vehicles, passengers, groups);
    }

    private FlowEdge createEdge(int from, int to, int capacity, BigDecimal cost) {
        FlowEdge edge = new FlowEdge();
        edge.setFrom(from);
        edge.setTo(to);
        edge.setCapacity(capacity);
        edge.setCost(cost);
        edge.setFlow(0);

        FlowEdge reverse = new FlowEdge();
        reverse.setFrom(to);
        reverse.setTo(from);
        reverse.setCapacity(0);
        reverse.setCost(cost.negate());
        reverse.setFlow(0);

        edge.setReverse(reverse);
        reverse.setReverse(edge);

        return edge;
    }

    private BigDecimal calculateCost(Vehicle vehicle, Passenger passenger) {
        double distance = calculateDistance(
                vehicle.getPickupLatitude(), vehicle.getPickupLongitude(),
                passenger.getPickupLatitude(), passenger.getPickupLongitude());

        double dropoffDistance = calculateDistance(
                passenger.getPickupLatitude(), passenger.getPickupLongitude(),
                passenger.getDropoffLatitude(), passenger.getDropoffLongitude());

        double totalDistance = distance + dropoffDistance;
        BigDecimal costPerKm = vehicle.getCostPerKm() != null ? vehicle.getCostPerKm() : BigDecimal.ZERO;
        return BigDecimal.valueOf(totalDistance).multiply(costPerKm);
    }

    private BigDecimal calculateGroupCost(Vehicle vehicle, PassengerGroup group) {
        List<Passenger> groupPassengers = group.getPassengers();
        if (groupPassengers == null || groupPassengers.isEmpty()) return BigDecimal.valueOf(Double.MAX_VALUE);

        BigDecimal totalCost = BigDecimal.ZERO;
        for (Passenger passenger : groupPassengers) {
            totalCost = totalCost.add(calculateCost(vehicle, passenger));
        }

        return totalCost;
    }

    private double calculateDistance(double lat1, double lon1, double lat2, double lon2) {
        final double R = 6371;
        double dLat = Math.toRadians(lat2 - lat1);
        double dLon = Math.toRadians(lon2 - lon1);
        double a = Math.sin(dLat / 2) * Math.sin(dLat / 2) +
                   Math.cos(Math.toRadians(lat1)) * Math.cos(Math.toRadians(lat2)) *
                   Math.sin(dLon / 2) * Math.sin(dLon / 2);
        double c = 2 * Math.atan2(Math.sqrt(a), Math.sqrt(1 - a));
        return R * c;
    }

    private FlowResult minCostMaxFlow(
            Map<Integer, List<FlowEdge>> graph,
            int source, int sink,
            int vehicleStart, int passengerStart, int groupStart,
            List<Vehicle> vehicles, List<Passenger> passengers, List<PassengerGroup> groups) {

        FlowResult result = new FlowResult();
        Map<Integer, List<Integer>> assignments = new HashMap<>();

        int maxIterations = 1000;
        for (int iteration = 0; iteration < maxIterations; iteration++) {
            Map<Integer, BigDecimal> distances = new HashMap<>();
            Map<Integer, FlowEdge> parents = new HashMap<>();
            Set<Integer> inQueue = new HashSet<>();

            for (int i = 0; i < graph.size(); i++) {
                distances.put(i, i == source ? BigDecimal.ZERO : BigDecimal.valueOf(Double.MAX_VALUE));
            }

            Queue<Integer> queue = new LinkedList<>();
            queue.add(source);
            inQueue.add(source);

            while (!queue.isEmpty()) {
                int u = queue.poll();
                inQueue.remove(u);

                for (FlowEdge edge : graph.get(u)) {
                    if (edge.getCapacity() - edge.getFlow() > 0) {
                        BigDecimal newDist = distances.get(u).add(edge.getCost());
                        if (newDist.compareTo(distances.get(edge.getTo())) < 0) {
                            distances.put(edge.getTo(), newDist);
                            parents.put(edge.getTo(), edge);

                            if (!inQueue.contains(edge.getTo())) {
                                queue.add(edge.getTo());
                                inQueue.add(edge.getTo());
                            }
                        }
                    }
                }
            }

            if (distances.get(sink).compareTo(BigDecimal.valueOf(Double.MAX_VALUE - 1000)) >= 0)
                break;

            List<FlowEdge> path = new ArrayList<>();
            int current = sink;
            int minFlow = Integer.MAX_VALUE;

            while (current != source) {
                if (!parents.containsKey(current))
                    break;

                FlowEdge edge = parents.get(current);
                path.add(edge);
                minFlow = Math.min(minFlow, edge.getCapacity() - edge.getFlow());
                current = edge.getFrom();
            }

            if (path.isEmpty())
                break;

            for (FlowEdge edge : path) {
                edge.setFlow(edge.getFlow() + minFlow);
                edge.getReverse().setFlow(edge.getReverse().getFlow() - minFlow);
            }

            result.setMaxFlow(result.getMaxFlow() + minFlow);
            result.setMinCost(result.getMinCost().add(distances.get(sink).multiply(BigDecimal.valueOf(minFlow))));

            for (FlowEdge edge : path) {
                if (edge.getFrom() >= vehicleStart && edge.getFrom() < passengerStart) {
                    int vehicleIndex = edge.getFrom() - vehicleStart;
                    int vehicleId = vehicles.get(vehicleIndex).getId();

                    if (edge.getTo() >= passengerStart && edge.getTo() < groupStart) {
                        int passengerIndex = edge.getTo() - passengerStart;
                        int passengerId = passengers.get(passengerIndex).getId();

                        assignments.computeIfAbsent(vehicleId, k -> new ArrayList<>()).add(passengerId);
                    } else if (edge.getTo() >= groupStart) {
                        int groupIndex = edge.getTo() - groupStart;
                        PassengerGroup group = groups.get(groupIndex);
                        List<Passenger> groupPassengers = passengers.stream()
                                .filter(p -> Objects.equals(p.getGroupId(), group.getId()))
                                .collect(Collectors.toList());

                        assignments.computeIfAbsent(vehicleId, k -> new ArrayList<>());
                        for (Passenger p : groupPassengers) {
                            assignments.get(vehicleId).add(p.getId());
                        }
                    }
                }
            }
        }

        result.setAssignments(assignments);
        return result;
    }
}

File: DoAnChuyenNganh-main/DoAnCoSo/WebDuLichDaLat/Services/MinCostMaxFlowService.cs
using WebDuLichDaLat.Models;
using System.Collections.Generic;
using System.Linq;

namespace WebDuLichDaLat.Services
{
    public class MinCostMaxFlowService
    {
        public class FlowEdge
        {
            public int From { get; set; }
            public int To { get; set; }
            public int Capacity { get; set; }
            public decimal Cost { get; set; }
            public int Flow { get; set; }
            public FlowEdge Reverse { get; set; }
        }

        public class FlowResult
        {
            public int MaxFlow { get; set; }
            public decimal MinCost { get; set; }
            public Dictionary<int, List<int>> Assignments { get; set; } = new Dictionary<int, List<int>>(); // VehicleId -> List of PassengerIds
        }

        public FlowResult Solve(List<Vehicle> vehicles, List<Passenger> passengers, List<PassengerGroup> groups)
        {
            // Tạo graph: Source -> Vehicles -> Passengers/Groups -> Sink
            int source = 0;
            int sink = 1;
            int vehicleStart = 2;
            int passengerStart = vehicleStart + vehicles.Count;
            int groupStart = passengerStart + passengers.Count;

            var graph = new Dictionary<int, List<FlowEdge>>();
            var allEdges = new List<FlowEdge>();

            // Khởi tạo graph (cần thêm sink node)
            int totalNodes = groupStart + groups.Count + 2; // +2 for source and sink
            for (int i = 0; i < totalNodes; i++)
            {
                graph[i] = new List<FlowEdge>();
            }

            // Source -> Vehicles
            for (int i = 0; i < vehicles.Count; i++)
            {
                int vehicleNode = vehicleStart + i;
                var edge = CreateEdge(source, vehicleNode, vehicles[i].AvailableSeats, 0);
                graph[source].Add(edge);
                graph[vehicleNode].Add(edge.Reverse);
                allEdges.Add(edge);
            }

            // Vehicles -> Passengers (individual)
            for (int v = 0; v < vehicles.Count; v++)
            {
                int vehicleNode = vehicleStart + v;
                var vehicle = vehicles[v];

                for (int p = 0; p < passengers.Count; p++)
                {
                    var passenger = passengers[p];
                    if (passenger.GroupId.HasValue) continue; // Bỏ qua nếu thuộc group

                    int passengerNode = passengerStart + p;
                    decimal cost = CalculateCost(vehicle, passenger);
                    var edge = CreateEdge(vehicleNode, passengerNode, 1, cost);
                    graph[vehicleNode].Add(edge);
                    graph[passengerNode].Add(edge.Reverse);
                    allEdges.Add(edge);
                }
            }

            // Vehicles -> Groups
            for (int v = 0; v < vehicles.Count; v++)
            {
                int vehicleNode = vehicleStart + v;
                var vehicle = vehicles[v];

                for (int g = 0; g < groups.Count; g++)
                {
                    var group = groups[g];
                    if (vehicle.AvailableSeats < group.RequiredSeats) continue;

                    int groupNode = groupStart + g;
                    decimal cost = CalculateGroupCost(vehicle, group);
                    var edge = CreateEdge(vehicleNode, groupNode, 1, cost);
                    graph[vehicleNode].Add(edge);
                    graph[groupNode].Add(edge.Reverse);
                    allEdges.Add(edge);
                }
            }

            // Groups -> Group Passengers
            for (int g = 0; g < groups.Count; g++)
            {
                int groupNode = groupStart + g;
                var group = groups[g];
                var groupPassengers = passengers.Where(p => p.GroupId == group.Id).ToList();

                foreach (var passenger in groupPassengers)
                {
                    int passengerNode = passengerStart + passengers.IndexOf(passenger);
                    var edge = CreateEdge(groupNode, passengerNode, 1, 0);
                    graph[groupNode].Add(edge);
                    graph[passengerNode].Add(edge.Reverse);
                    allEdges.Add(edge);
                }
            }

            // Passengers -> Sink
            for (int p = 0; p < passengers.Count; p++)
            {
                int passengerNode = passengerStart + p;
                var edge = CreateEdge(passengerNode, sink, 1, 0);
                graph[passengerNode].Add(edge);
                graph[sink].Add(edge.Reverse);
                allEdges.Add(edge);
            }

            // Chạy Min-Cost Max-Flow (Bellman-Ford + DFS)
            return MinCostMaxFlow(graph, source, sink, vehicleStart, passengerStart, groupStart, vehicles, passengers, groups);
        }

        private FlowEdge CreateEdge(int from, int to, int capacity, decimal cost)
        {
            var edge = new FlowEdge
            {
                From = from,
                To = to,
                Capacity = capacity,
                Cost = cost,
                Flow = 0
            };

            var reverse = new FlowEdge
            {
                From = to,
                To = from,
                Capacity = 0,
                Cost = -cost,
                Flow = 0
            };

            edge.Reverse = reverse;
            reverse.Reverse = edge;

            return edge;
        }

        private decimal CalculateCost(Vehicle vehicle, Passenger passenger)
        {
            // Tính chi phí dựa trên khoảng cách
            double distance = CalculateDistance(
                vehicle.PickupLatitude, vehicle.PickupLongitude,
                passenger.PickupLatitude, passenger.PickupLongitude);

            double dropoffDistance = CalculateDistance(
                passenger.PickupLatitude, passenger.PickupLongitude,
                passenger.DropoffLatitude, passenger.DropoffLongitude);

            double totalDistance = distance + dropoffDistance;
            return (decimal)totalDistance * vehicle.CostPerKm;
        }

        private decimal CalculateGroupCost(Vehicle vehicle, PassengerGroup group)
        {
            // Tính chi phí cho cả nhóm (có thể tối ưu hơn)
            var groupPassengers = group.Passengers.ToList();
            if (!groupPassengers.Any()) return decimal.MaxValue;

            decimal totalCost = 0;
            foreach (var passenger in groupPassengers)
            {
                totalCost += CalculateCost(vehicle, passenger);
            }

            return totalCost;
        }

        private double CalculateDistance(double lat1, double lon1, double lat2, double lon2)
        {
            const double R = 6371;
            double dLat = ToRadians(lat2 - lat1);
            double dLon = ToRadians(lon2 - lon1);
            double a = Math.Sin(dLat / 2) * Math.Sin(dLat / 2) +
                       Math.Cos(ToRadians(lat1)) * Math.Cos(ToRadians(lat2)) *
                       Math.Sin(dLon / 2) * Math.Sin(dLon / 2);
            double c = 2 * Math.Atan2(Math.Sqrt(a), Math.Sqrt(1 - a));
            return R * c;
        }

        private double ToRadians(double angle) => angle * Math.PI / 180;

        private FlowResult MinCostMaxFlow(
            Dictionary<int, List<FlowEdge>> graph,
            int source, int sink,
            int vehicleStart, int passengerStart, int groupStart,
            List<Vehicle> vehicles, List<Passenger> passengers, List<PassengerGroup> groups)
        {
            var result = new FlowResult();
            var assignments = new Dictionary<int, List<int>>();

            // Sử dụng Bellman-Ford để tìm đường đi có chi phí tối thiểu
            int maxIterations = 1000;
            for (int iteration = 0; iteration < maxIterations; iteration++)
            {
                var distances = new Dictionary<int, decimal>();
                var parents = new Dictionary<int, FlowEdge>();
                var inQueue = new HashSet<int>();

                for (int i = 0; i < graph.Count; i++)
                {
                    distances[i] = i == source ? 0 : decimal.MaxValue;
                }

                var queue = new Queue<int>();
                queue.Enqueue(source);
                inQueue.Add(source);

                while (queue.Count > 0)
                {
                    int u = queue.Dequeue();
                    inQueue.Remove(u);

                    foreach (var edge in graph[u])
                    {
                        if (edge.Capacity - edge.Flow > 0)
                        {
                            decimal newDist = distances[u] + edge.Cost;
                            if (newDist < distances[edge.To])
                            {
                                distances[edge.To] = newDist;
                                parents[edge.To] = edge;

                                if (!inQueue.Contains(edge.To))
                                {
                                    queue.Enqueue(edge.To);
                                    inQueue.Add(edge.To);
                                }
                            }
                        }
                    }
                }

                if (distances[sink] == decimal.MaxValue)
                    break;

                // Tìm đường đi từ source đến sink
                var path = new List<FlowEdge>();
                int current = sink;
                int minFlow = int.MaxValue;

                while (current != source)
                {
                    if (!parents.ContainsKey(current))
                        break;

                    var edge = parents[current];
                    path.Add(edge);
                    minFlow = Math.Min(minFlow, edge.Capacity - edge.Flow);
                    current = edge.From;
                }

                if (path.Count == 0)
                    break;

                // Cập nhật flow
                foreach (var edge in path)
                {
                    edge.Flow += minFlow;
                    edge.Reverse.Flow -= minFlow;
                }

                result.MaxFlow += minFlow;
                result.MinCost += distances[sink] * minFlow;

                // Cập nhật assignments
                foreach (var edge in path)
                {
                    if (edge.From >= vehicleStart && edge.From < passengerStart)
                    {
                        int vehicleIndex = edge.From - vehicleStart;
                        int vehicleId = vehicles[vehicleIndex].Id;

                        if (edge.To >= passengerStart && edge.To < groupStart)
                        {
                            int passengerIndex = edge.To - passengerStart;
                            int passengerId = passengers[passengerIndex].Id;

                            if (!assignments.ContainsKey(vehicleId))
                                assignments[vehicleId] = new List<int>();
                            assignments[vehicleId].Add(passengerId);
                        }
                        else if (edge.To >= groupStart)
                        {
                            int groupIndex = edge.To - groupStart;
                            var group = groups[groupIndex];
                            var groupPassengers = passengers.Where(p => p.GroupId == group.Id).ToList();

                            if (!assignments.ContainsKey(vehicleId))
                                assignments[vehicleId] = new List<int>();
                            foreach (var p in groupPassengers)
                            {
                                assignments[vehicleId].Add(p.Id);
                            }
                        }
                    }
                }
            }

            result.Assignments = assignments;
            return result;
        }
    }
}





